t/08-l1_serializer.t

```diff
@@ -0,0 +1,417 @@
+            local cache, err = mlcache.new("my_mlcache", "cache_shm")
+
+            if not cache then
+                ngx.log(ngx.ERR, err)
+                return
+            end
+
+            local data, err = cache:get("key", {
+                l1_serializer = function(s)
+                    return string.format("transform(%q)", s)
+                end
+            }, function() return "foo" end)
+            if not data then
+                ngx.say(ngx.ERR, err)
+            end
+            ngx.say(data)
+        }
+    }
+--- request
+GET /t
+--- response_body
+transform("foo")
+--- no_error_log
+[error]
+
+
+
+=== TEST 9: l1_serializer as a :get() parameter has precedence over the constructor one
+--- http_config eval: $::HttpConfig
+--- config
+    location = /t {
+        content_by_lua_block {
+            local mlcache = require "resty.mlcache"
+
+            local cache, err = mlcache.new("my_mlcache", "cache_shm", {
+                l1_serializer = function(s)
+                    return string.format("constructor(%q)", s)
+                end
+            })
+
+            if not cache then
+                ngx.log(ngx.ERR, err)
+                return
+            end
+
+            local data, err = cache:get("key1", {
+                l1_serializer = function(s)
+                    return string.format("get_parameter(%q)", s)
+                end
+            }, function() return "foo" end)
+            if not data then
+                ngx.say(ngx.ERR, err)
+            end
+            ngx.say(data)
+
+            local data, err = cache:get("key2", nil, function() return "bar" end)
+            if not data then
+                ngx.say(ngx.ERR, err)
+            end
+            ngx.say(data)
+        }
+    }
+--- request
+GET /t
+--- response_body
+get_parameter("foo")
+constructor("bar")
+--- no_error_log
+[error]
+
+
+
+=== TEST 10: l1_serializer is called for set calls
+--- http_config eval: $::HttpConfig
+--- config
+    location = /t {
+        content_by_lua_block {
+            local mlcache = require "resty.mlcache"
+
+            local cache, err = mlcache.new("my_mlcache", "cache_shm", {
+                ipc_shm = "ipc_shm",
+            })
+
+            if not cache then
+                ngx.log(ngx.ERR, err)
+                return
+            end
+
+            local called = false
+            local ok, err = cache:set("key", {
+                l1_serializer = function(s)
+                    called = true
+                    return string.format("transform(%q)", s)
+                end
+            }, "value")
+
+            if not ok then
+                ngx.say(err)
+            end
+            ngx.say(tostring(called))
+
+            local value, err = cache:get("key", nil, error)
+            if not value then
+                ngx.say(err)
+            end
+            ngx.say(value)
+        }
+    }
+--- request
+GET /t
+--- response_body
+true
+transform("value")
+--- no_error_log
+[error]
```

Comment:
Maybe we are missing another test: assert that the `l1_serializer` is not called when L2 hit returns a cached miss (upon negative caching, basically). This way we assert our documentation when it says "the argument to `l1_serializer` can never be `nil`"